src/types/mod.rs

```diff
@@ -172,6 +180,43 @@ macro_rules! pyobject_native_type_base(
 // make sense on PyAny / have different implementations).
 #[doc(hidden)]
 #[macro_export]
+#[cfg_attr(docsrs, doc(cfg(all())))]
+#[cfg(not(feature = "gil-refs"))]
+macro_rules! pyobject_native_type_named (
+    ($name:ty $(;$generics:ident)*) => {
+
+        impl<$($generics,)*> ::std::convert::AsRef<$crate::PyAny> for $name {
+            #[inline]
+            fn as_ref(&self) -> &$crate::PyAny {
+                &self.0
+            }
+        }
+
+        impl<$($generics,)*> ::std::ops::Deref for $name {
+            type Target = $crate::PyAny;
+
+            #[inline]
+            fn deref(&self) -> &$crate::PyAny {
+                &self.0
+            }
+        }
```

Comment:
I don't think so, in general I think it's helpful to not have the structure of the type objects be constrained by `AsRef` / `Deref`.

Reply:
Oh I see what you mean, I've duplicated too much here. Will fix later!

Reply:
Actually we still have `AsRef` / `Deref` here in this macro on `main`, maybe we could remove later (i.e. in 0.23 or later) but I think we have to leave these here in this PR for the patch fix. `AsPyPointer` (just below) should be removed though, I pushed that.

Reply:
Sure, removing them from `main` makes sense. I don't think these can ever be used now, because there is no way to construct a literal `PyAny` (or similar)